freqtrade/freqai/utils.py

```diff
@@ -85,6 +93,83 @@ def get_required_data_timerange(
+        informative = self.dp.get_pair_dataframe(pair, tf)
+
+    # first loop is automatically duplicating indicators for time periods
+    for t in self.freqai_info["feature_parameters"]["indicator_periods_candles"]:
+
+        t = int(t)
+        informative[f"%-{coin}rsi-period_{t}"] = ta.RSI(informative, timeperiod=t)
+        informative[f"%-{coin}mfi-period_{t}"] = ta.MFI(informative, timeperiod=t)
+        informative[f"%-{coin}adx-period_{t}"] = ta.ADX(informative, timeperiod=t)
+        informative[f"%-{coin}sma-period_{t}"] = ta.SMA(informative, timeperiod=t)
+        informative[f"%-{coin}ema-period_{t}"] = ta.EMA(informative, timeperiod=t)
+
+        bollinger = qtpylib.bollinger_bands(
+            qtpylib.typical_price(informative), window=t, stds=2.2
+        )
+        informative[f"{coin}bb_lowerband-period_{t}"] = bollinger["lower"]
+        informative[f"{coin}bb_middleband-period_{t}"] = bollinger["mid"]
+        informative[f"{coin}bb_upperband-period_{t}"] = bollinger["upper"]
+
+        informative[f"%-{coin}bb_width-period_{t}"] = (
+            informative[f"{coin}bb_upperband-period_{t}"]
+            - informative[f"{coin}bb_lowerband-period_{t}"]
+        ) / informative[f"{coin}bb_middleband-period_{t}"]
+        informative[f"%-{coin}close-bb_lower-period_{t}"] = (
+            informative["close"] / informative[f"{coin}bb_lowerband-period_{t}"]
+        )
+
+        informative[f"%-{coin}roc-period_{t}"] = ta.ROC(informative, timeperiod=t)
+
+        informative[f"%-{coin}relative_volume-period_{t}"] = (
+            informative["volume"] / informative["volume"].rolling(t).mean()
+        )
+
+    informative[f"%-{coin}pct-change"] = informative["close"].pct_change()
+    informative[f"%-{coin}raw_volume"] = informative["volume"]
+    informative[f"%-{coin}raw_price"] = informative["close"]
+
+    indicators = [col for col in informative if col.startswith("%")]
+    # This loop duplicates and shifts all indicators to add a sense of recency to data
+    for n in range(self.freqai_info["feature_parameters"]["include_shifted_candles"] + 1):
+        if n == 0:
+            continue
+        informative_shift = informative[indicators].shift(n)
+        informative_shift = informative_shift.add_suffix("_shift-" + str(n))
+        informative = pd.concat((informative, informative_shift), axis=1)
+
+    df = merge_informative_pair(df, informative, self.config["timeframe"], tf, ffill=True)
+    skip_columns = [
+        (s + "_" + tf) for s in ["date", "open", "high", "low", "close", "volume"]
+    ]
+    df = df.drop(columns=skip_columns)
+    if set_generalized_indicators:
+        df["%-day_of_week"] = (df["date"].dt.dayofweek + 1) / 7
+        df["%-hour_of_day"] = (df["date"].dt.hour + 1) / 25
+        df["&s-minima"] = 0
+        df["&s-maxima"] = 0
+        min_peaks = argrelextrema(df["close"].values, np.less, order=80)
+        max_peaks = argrelextrema(df["close"].values, np.greater, order=80)
```

Comment:
are you sure this is working here?

iirc (been a while since i looked at this though), this doesn't calculate min/max values in a rolling fashion - but is using the full dataframe at any time - therefore causing an implicit lookahead (and with that - bias in backtesting).

Reply:
It's important to remember that FreqAI targets *need* to look into the future. 

Nonetheless, live/backtesting do have slightly different behaviors (not hugely). For live mode, FreqAI trims the data frame to the training period and feeds it into `populate_any_indicators`

In backtesting,full df gets maxmins from `populate_any`, then the training gets trimmed to the window and we train on those labels. Thus the edge labels of each window are "more accurate". But that edge is no more than 1 label on each side. Typically users are training on thousands of labels per train window so this accounts for less than 1%.

We could try to make backtesting emulate live closer by slicing data frame and then `populate_any` For each window - It may need to be a separate PR though depending on ramifications.

Reply:
i'm just ... asking if we're sure it's working here - not saying it's not ;) 

> We could try to make backtesting emulate live closer by slicing data frame and then populate_any For each window - It may need to be a separate PR though depending on ramifications.

I've tried this once (for each candle, actually) - and the performance hit is VERY high - taking you from a 10s backtest to 5-6 minutes easily.

Reply:
I don't think it would in our case - it all happens inside the `IFreqaiModel` the strategy/backtest sends the full dataframe in and gets the full dataframe out. IT is just  about *when* freqai slices it. I will likely change this in a separate PR because it is objectively better and closer to live.